File: src/main/org/nlogo/mirror/PatchData.scala
// (C) 2012 Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.mirror

import org.nlogo.api
import org.nlogo.api.LogoList
import org.nlogo.api.Color.{ getARGBIntByRGBAList, getRGBAListByARGB }
import java.io.{ DataInputStream, DataOutputStream, IOException }

object PatchData {

  val DEAD         = 0x0000
  val PCOLOR       = 0x0001
  val PLABEL       = 0x0002
  val PLABEL_COLOR = 0x0004
  val PXCOR        = 0x0008
  val PYCOR        = 0x0010
  val COMPLETE = PCOLOR | PLABEL | PLABEL_COLOR | PXCOR | PYCOR

  val OverrideVariables = Array("PCOLOR", "PLABEL", "PLABEL-COLOR")
  val OverrideMethods = Array("pcolor", "plabel", "plabelColor")
  
  def getOverrideIndex(varName: String): Int =
    Overridable.getOverrideIndex(OverrideVariables, varName)

  @throws(classOf[java.io.IOException])
  def fromStream(in: DataInputStream) = {
    val id = in.readLong()
    val mask = in.readShort()
    new PatchData(
      id, mask,
      if ((mask & PXCOR) == PXCOR)
        in.readInt() else 0,
      if ((mask & PYCOR) == PYCOR)
        in.readInt() else 0,
      if ((mask & PCOLOR) == PCOLOR)
        getRGBAListByARGB(in.readInt()) else null,
      if ((mask & PLABEL) == PLABEL)
        in.readUTF() else null,
      if ((mask & PLABEL_COLOR) == PLABEL_COLOR)
        getRGBAListByARGB(in.readInt()) else null)
  }

}

/**
 * The mask specifies which patch attributes are described by this PatchData object, as the bitwise
 * or of all the description masks that apply to this PatchData.
 *
 * For example: a PatchData that describes the color and the label color of a patch, but not the
 * label itself should have mask <code>PCOLOR | LABEL_COLOR</code>.
 */
class PatchData private (val id: Long, var mask: Int, var pxcor: Int, var pycor: Int, var _pcolor: LogoList, var plabel: String, var _plabelColor: LogoList, var patchColors: Array[Int] = null)
extends AgentData {

  def this(id: Long, mask: Int, pxcor: Int, pycor: Int, pcolor: AnyRef, plabel: String, plabelColor: AnyRef) =
    this(id, mask, pxcor, pycor, AgentData.toLogoList(pcolor), plabel, AgentData.toLogoList(plabelColor), null)

  def this(id: Long) =
    this(id, PatchData.DEAD, 0, 0, null, null, null, null)

  import PatchData._

  def plabel_=(_label: AnyRef) {
    plabel = api.Dump.logoObject(_label)
  }

  def hasLabel = plabel.nonEmpty

  def methodName(index: Int) =
    PatchData.OverrideMethods(index)

  def pcolor = _pcolor
  def pcolor_=(pc: AnyRef) {
    _pcolor = AgentData.toLogoList(pc)
    if (patchColors != null)
      patchColors(id.toInt) = getARGBIntByRGBAList(_pcolor)
  }

  def plabelColor = _plabelColor
  def plabelColor_=(plc: AnyRef) { _plabelColor = AgentData.toLogoList(plc) }

  override def xcor = pxcor
  override def ycor = pycor
  override def spotlightSize = 1
  override def wrapSpotlight = false

  def stringRep =
    "Patch " + pxcor + " " + pycor + " (" + getARGBIntByRGBAList(pcolor) +
    ", " + plabel + ", " + getARGBIntByRGBAList(plabelColor) + ")"

  def isComplete =
    (mask & COMPLETE) == COMPLETE

  /**
   * Updates this PatchData with data from another PatchData.
   *
   * @param otherPatch another PatchData from which to update this PatchData
   * @return a new PatchData representing the changes made to
   *         this PatchData (i.e. the difference between the two objects).
   *         If nothing was changed, returns null.
   */
  def updateFrom(otherPatch: PatchData): PatchData = {
    // start out with a "dead" patch, but we'll fill it in...
    val diffs = new PatchData(id)
    if ((otherPatch.mask & PXCOR) == PXCOR && pxcor != otherPatch.pxcor) {
      pxcor = otherPatch.pxcor
      diffs.mask |= PXCOR
      diffs.pxcor = pxcor
    }
    if ((otherPatch.mask & PYCOR) == PYCOR && pycor != otherPatch.pycor) {
      pycor = otherPatch.pycor
      diffs.mask |= PYCOR
      diffs.pycor = pycor
    }
    if ((otherPatch.mask & PCOLOR) == PCOLOR && !pcolor.equals(otherPatch.pcolor)) {
      _pcolor = otherPatch.pcolor
      diffs.mask |= PCOLOR
      diffs._pcolor = pcolor
    }
    if ((otherPatch.mask & PLABEL) == PLABEL && !plabel.equals(otherPatch.plabel)) {
      plabel = otherPatch.plabel
      diffs.mask |= PLABEL
      diffs.plabel = plabel
    }
    if ((otherPatch.mask & PLABEL_COLOR) == PLABEL_COLOR && !plabelColor.equals(otherPatch.plabelColor)) {
      plabelColor = otherPatch.plabelColor
      diffs.mask |= PLABEL_COLOR
      diffs.plabelColor = plabelColor
    }
    if (diffs.mask != DEAD)
      diffs else null
  }

  override def serialize(out: DataOutputStream) {
    out.writeLong(id)
    out.writeShort(mask.toShort)
    if ((mask & PXCOR) == PXCOR)
      out.writeInt(pxcor)
    if ((mask & PYCOR) == PYCOR)
      out.writeInt(pycor)
    if ((mask & PCOLOR) == PCOLOR)
      out.writeInt(getARGBIntByRGBAList(pcolor))
    if ((mask & PLABEL) == PLABEL)
      out.writeUTF(plabel)
    if ((mask & PLABEL_COLOR) == PLABEL_COLOR)
      out.writeInt(api.Color.getARGBIntByRGBAList(plabelColor))
  }
}

File: headless/src/main/org/nlogo/hubnet/mirroring/PatchData.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.hubnet.mirroring;

import org.nlogo.api.Dump;
import org.nlogo.api.LogoList;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

/**
 * Describes a single patch.
 */
public final strictfp class PatchData
    extends AgentData {

  /**
   * Description mask for PatchData object with no descriptive content. *
   */
  public static final short DEAD = 0x0000;

  /**
   * Description mask for patch pcolor. *
   */
  public static final short PCOLOR = 0x0001;

  /**
   * Description mask for patch label. *
   */
  public static final short PLABEL = 0x0002;

  /**
   * Description mask for label color. *
   */
  public static final short PLABEL_COLOR = 0x0004;

  public static final short PXCOR = 0x0008;
  public static final short PYCOR = 0x0010;

  /**
   * Description mask for PatchData object that completely describes a patch. *
   */
  public static final short COMPLETE = PCOLOR | PLABEL | PLABEL_COLOR | PXCOR | PYCOR;

  /**
   * The id of the patch. *
   */
  private final long id;

  /**
   * The color of the patch. *
   */
  private LogoList pcolor;

  /**
   * The label of the patch. *
   */
  private String plabel = "";

  /**
   * The label color of the patch. *
   */
  private LogoList plabelColor;

  private int pxcor;
  private int pycor;

  /**
   * Specifies which patch attributes are described by this
   * PatchData object.
   * The bitwise or of all the description masks that apply
   * to this PatchData.<p>
   * For example: a PatchData that describes the color and
   * the label color of a patch, but not the label itself
   * should have mask <code>PCOLOR | LABEL_COLOR</code>.
   */
  private short mask;

  int[] patchColors;


  /**
   * Creates a new patch data.
   */
  PatchData(long id, short mask, int pxcor, int pycor, Object pcolor, String plabel, Object plabelColor) {
    this.id = id;
    this.pxcor = pxcor;
    this.pycor = pycor;
    this.mask = mask;
    pcolor(pcolor);
    this.plabel = plabel;
    plabelColor(plabelColor);
  }

  /**
   * Reconstructs a serialized patch data object.
   *
   * @param is stream containing serialized patch data.
   */
  PatchData(DataInputStream is)
      throws IOException {
    id = is.readLong();
    mask = is.readShort();
    if ((mask & PXCOR) == PXCOR) {
      pxcor = is.readInt();
    }
    if ((mask & PYCOR) == PYCOR) {
      pycor = is.readInt();
    }
    if ((mask & PCOLOR) == PCOLOR) {
      pcolor(is.readInt());
    }
    if ((mask & PLABEL) == PLABEL) {
      plabel = is.readUTF();
    }
    if ((mask & PLABEL_COLOR) == PLABEL_COLOR) {
      plabelColor(is.readInt());
    }
  }

  /**
   * Creates a new patch data with no descriptive content.
   */
  private PatchData(long id) {
    this.id = id;
    mask = DEAD;
  }

  /**
   * Returns the id of the patch this object describes.
   */
  public long id() {
    return id;
  }

  public int pxcor() {
    return pxcor;
  }

  public int pycor() {
    return pycor;
  }

  /**
   * Returns the color of the patch this object describes.
   */
  public Object pcolor() {
    return pcolor;
  }

  public void pcolor(Object pcolor) {
    if (pcolor instanceof Double) {
      pcolor((Double) pcolor);
    } else {
      pcolor((LogoList) pcolor);
    }
  }

  private void pcolor(Double pcolor) {
    pcolor(org.nlogo.api.Color.getARGBbyPremodulatedColorNumber(pcolor));
  }

  private void pcolor(org.nlogo.api.LogoList pcolor) {
    this.pcolor = pcolor;
    if (patchColors != null) {
      patchColors[(int) id] = org.nlogo.api.Color.getARGBIntByRGBAList(pcolor);
    }
  }

  private void pcolor(int argb) {
    this.pcolor = org.nlogo.api.Color.getRGBAListByARGB(argb);
    if (patchColors != null) {
      patchColors[(int) id] = argb;
    }
  }

  /**
   * Returns the label of the patch this object describes.
   */
  public String plabel() {
    return plabel;
  }

  public void plabel(Object plabel) {
    this.plabel = Dump.logoObject(plabel);
  }

  /**
   * Returns the label color of the patch this object describes.
   */
  public Object plabelColor() {
    return plabelColor;
  }

  public void plabelColor(Object plabelColor) {
    if (plabelColor instanceof Double) {
      plabelColor((Double) plabelColor);
    } else {
      plabelColor((LogoList) plabelColor);
    }
  }

  public void plabelColor(Double plabelColor) {
    plabelColor(org.nlogo.api.Color.getARGBbyPremodulatedColorNumber(plabelColor));
  }

  public void plabelColor(LogoList plabelColor) {
    this.plabelColor = plabelColor;
  }

  public void plabelColor(int argb) {
    plabelColor(org.nlogo.api.Color.getRGBAListByARGB(argb));
  }

  public boolean hasLabel() {
    return plabel.length() > 0;
  }

  private static final String[] OVERRIDE_VARIABLES = new String[]{"PCOLOR", "PLABEL", "PLABEL-COLOR"};
  private static final String[] OVERRIDE_METHODS = new String[]{"pcolor", "plabel", "plabelColor"};

  @Override
  public String getMethodName(int index) {
    return OVERRIDE_METHODS[index];
  }

  public static int getOverrideIndex(String varName) {
    return getOverrideIndex(OVERRIDE_VARIABLES, varName);
  }

  @Override
  public double xcor() {
    return pxcor;
  }

  @Override
  public double ycor() {
    return pycor;
  }

  @Override
  public double spotlightSize() {
    return 1;
  }

  @Override
  public boolean wrapSpotlight() {
    return false;
  }

  /**
   * Returns a string representation of the patch this object describes.
   */
  String stringRep() {
    return "Patch " + pxcor + " " + pycor + " (" + org.nlogo.api.Color.getARGBIntByRGBAList(pcolor)
        + ", " + plabel + ", " + org.nlogo.api.Color.getARGBIntByRGBAList(plabelColor) + ")";
  }

  /**
   * Returns true if this PatchData completely describes a patch.
   */
  boolean isComplete() {
    return (mask & COMPLETE) == COMPLETE;
  }

  /**
   * Updates this PatchData with data from another PatchData.
   *
   * @param otherPatch another PatchData from which to update this PatchData
   * @return a new PatchData representing the changes made to
   *         this PatchData (i.e. the difference between the two objects).
   *         If nothing was changed, returns null.
   */
  PatchData updateFrom(PatchData otherPatch) {
    // start out with a "dead" patch, but we'll fill it in...
    PatchData diffs = new PatchData(id);

    // update the values...
    if ((otherPatch.mask & PXCOR) == PXCOR && pxcor != otherPatch.pxcor) {
      pxcor = otherPatch.pxcor;
      diffs.mask |= PXCOR;
      diffs.pxcor = pxcor;
    }
    if ((otherPatch.mask & PYCOR) == PYCOR && pycor != otherPatch.pycor) {
      pycor = otherPatch.pycor;
      diffs.mask |= PYCOR;
      diffs.pycor = pycor;
    }
    if ((otherPatch.mask & PCOLOR) == PCOLOR && !pcolor.equals(otherPatch.pcolor)) {
      pcolor = otherPatch.pcolor;
      diffs.mask |= PCOLOR;
      diffs.pcolor = pcolor;
    }
    if ((otherPatch.mask & PLABEL) == PLABEL && !plabel.equals(otherPatch.plabel)) {
      plabel = otherPatch.plabel;
      diffs.mask |= PLABEL;
      diffs.plabel = plabel;
    }
    if ((otherPatch.mask & PLABEL_COLOR) == PLABEL_COLOR && !plabelColor.equals(otherPatch.plabelColor)) {
      plabelColor(otherPatch.plabelColor);
      diffs.mask |= PLABEL_COLOR;
      diffs.plabelColor = plabelColor;
    }

    if (diffs.mask != DEAD) {
      return diffs;
    } else {
      return null;
    }
  }

  /**
   * Serializes this patch data object.
   */
  void serialize(DataOutputStream os)
      throws IOException {
    os.writeLong(id);
    os.writeShort(mask);
    if ((mask & PXCOR) == PXCOR) {
      os.writeInt(pxcor);
    }
    if ((mask & PYCOR) == PYCOR) {
      os.writeInt(pycor);
    }
    if ((mask & PCOLOR) == PCOLOR) {
      os.writeInt(org.nlogo.api.Color.getARGBIntByRGBAList(pcolor));
    }
    if ((mask & PLABEL) == PLABEL) {
      os.writeUTF(plabel);
    }
    if ((mask & PLABEL_COLOR) == PLABEL_COLOR) {
      os.writeInt(org.nlogo.api.Color.getARGBIntByRGBAList(plabelColor));
    }
  }
}
